source/agora/protocol/GossipProtocol.d

```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+
+    Contains the code used for peer-to-peer communication.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.protocol.GossipProtocol;
+import agora.common.Data;
+import agora.protocol.Cache;
+
+alias MessageCache = Cache!(Hash, bool);
+
+/// Interface of Protocol
+interface IProtocol {
+    public bool receiveMessage(Hash msg);
+}
+
+/// Interface of Protocol Owner 
+interface IProtocolOwner {
+    public bool sendMessage(Hash msg);
+
+}
+
+/// Procedure of peer-to-peer communication
+class GossipProtocol : IProtocol {
+    private IProtocolOwner owner;
+    private MessageCache receivedMsgCache;
```

Comment:
That sounds like the `Set` @AndrejMitrovic implemented

Reply:
The comments seem to have been changed up and down.

Reply:
@Geod24 I could use some feedback in https://github.com/bpfkorea/agora/pull/14#issuecomment-504297575 before that `Set` is merged.

Or we could split that `Set` implementation out of the PR so it can be used here?

Reply:
Modified.

Reply:
@Geod24 Check Please